dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java

```diff
@@ -0,0 +1,215 @@
+
+    private ProgramInstance pi1;
+
+    private ProgramInstance pi2;
+
+    private ProgramInstance pi3;
+
+    private ProgramInstance pi4;
+
+    private TrackedEntityInstance tei1;
+
+    private TrackedEntityInstance tei2;
+
+    private TrackedEntityInstance tei3;
+
+    private TrackedEntityInstance tei4;
+
+    private TrackedEntityType teiType;
+
+    private User user;
+
+    @Override
+    protected void setUpTest()
+        throws Exception
+    {
+        userService = _userService;
+        user = createAndInjectAdminUser();
+
+        orgUnitA = createOrganisationUnit( "A" );
+        organisationUnitService.addOrganisationUnit( orgUnitA );
+
+        user.getOrganisationUnits().add( orgUnitA );
+
+        teiType = createTrackedEntityType( 'P' );
+        trackedEntityTypeService.addTrackedEntityType( teiType );
+
+        program = createProgram( 'P' );
+        programService.addProgram( program );
+
+        tei1 = createTrackedEntityInstance( orgUnitA );
+        tei2 = createTrackedEntityInstance( orgUnitA );
+        tei3 = createTrackedEntityInstance( orgUnitA );
+        tei4 = createTrackedEntityInstance( orgUnitA );
+        tei1.setTrackedEntityType( teiType );
+        tei2.setTrackedEntityType( teiType );
+        tei3.setTrackedEntityType( teiType );
+        tei4.setTrackedEntityType( teiType );
+
+        trackedEntityInstanceService.addTrackedEntityInstance( tei1 );
+        trackedEntityInstanceService.addTrackedEntityInstance( tei2 );
+        trackedEntityInstanceService.addTrackedEntityInstance( tei3 );
+        trackedEntityInstanceService.addTrackedEntityInstance( tei4 );
+
+        pi1 = createProgramInstance( program, tei1, orgUnitA );
+        pi2 = createProgramInstance( program, tei2, orgUnitA );
+        pi3 = createProgramInstance( program, tei3, orgUnitA );
+        pi4 = createProgramInstance( program, tei4, orgUnitA );
+
+        programInstanceService.addProgramInstance( pi1 );
+        programInstanceService.addProgramInstance( pi2 );
+        programInstanceService.addProgramInstance( pi3 );
+        programInstanceService.addProgramInstance( pi4 );
+
+        programInstanceService.enrollTrackedEntityInstance( tei1, program, new Date(), new Date(), orgUnitA );
+        programInstanceService.enrollTrackedEntityInstance( tei2, program, new Date(), new Date(), orgUnitA );
+        programInstanceService.enrollTrackedEntityInstance( tei3, program, new Date(), new Date(), orgUnitA );
+        programInstanceService.enrollTrackedEntityInstance( tei4, program, new Date(), new Date(), orgUnitA );
+
+        userService.addUser( user );
+    }
+
+    @Test
+    void testConfiguredMaxTeiLimit()
+    {
+        systemSettingManager.saveSystemSetting( SettingKey.TRACKED_ENTITY_MAX_LIMIT, 3 );
+        TrackedEntityInstanceQueryParams params = new TrackedEntityInstanceQueryParams();
+        params.setProgram( program );
+        params.setOrganisationUnits( Sets.newHashSet( orgUnitA ) );
```

Comment:
Please prefer Java's `Set.of` for something like that 😄 (mind https://github.com/dhis2/wow-backend/blob/master/guides/java11_pitfalls.md)